iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -492,13 +478,75 @@ static LogicalResult analyseScfForOp(scf::ForOp forOp,
   auto regionArgs = forOp.getRegionIterArgs();
   auto initArgs = forOp.initArgs();
   for (int i = 0; i < yeildOp.results().size(); ++i) {
-    Value outputTensor = yeildOp.results()[i];
+    Value yieldTensor = yeildOp.results()[i];
     Value resultTensor = forOp.results()[i];
     Value initArg = initArgs[i];
     Value arg = regionArgs[i];
-    plan.unionSets(outputTensor, resultTensor);
-    plan.unionSets(outputTensor, initArg);
-    plan.unionSets(outputTensor, arg);
+    // Always tie the yield, the result tensor, and the region arg
+    plan.unionSets(yieldTensor, resultTensor);
+    plan.unionSets(yieldTensor, arg);
+
+    // If the init value is not read-only and has single use, the tie the init
+    // and result (and by extension all 4 tensors here).
+    if (hasSingleRealUse(initArg) && !isFromReadOnlyTensor(initArg, plan)) {
+      plan.unionSets(initArg, resultTensor);
+    }
+  }
+  return success();
+}
+
+/// Look for destructive update loop pattern.
+///
+/// ```mlir
+///   %result = scf.for %arg0 = ... iter_args(%arg1 = %init) {
+///     %st = subtensor %arg1[...]
+///
+///     %yieldVal = subtensor_insert %val, %arg1[...]
+///     scf.yield %yieldVal
+///   }
+///
+/// `%result`, `%arg1` and `%yieldVal` are all already in the same equivalence
+/// class. `%st` and `%arg` can be added to the same equivalence class even
+/// though `%arg1` has multiple uses. Same is true for `%yieldVal` and
+/// `%arg1`. Here we also verify there are no other "value" uses of
+/// `%arg1`. This might be overly constraining, but we can relax gradually.
+static LogicalResult hasDestructiveUpdateLoopPattern(scf::ForOp forOp,
+                                                     BufferizationPlan &plan) {
+  for (BlockArgument arg : forOp.getRegionIterArgs()) {
+    bool allArgUsesCompatible = true;
+    for (OpOperand &use : arg.getUses()) {
+      Operation *user = use.getOwner();
+      bool isDestructiveUpdateCompatibleUse =
+          TypeSwitch<Operation *, bool>(user)
+              .Case<SubTensorOp>([&](SubTensorOp subTensorOp) {
+                return subTensorOp.source() == arg;
+              })
+              .Case<SubTensorInsertOp>(
+                  [&](SubTensorInsertOp subTensorInsertOp) {
+                    return subTensorInsertOp.dest() == arg;
+                  })
+              .Case<memref::DimOp, scf::YieldOp>([&](auto op) { return true; })
+              .Default([&](Operation *op) { return false; });
+      if (!isDestructiveUpdateCompatibleUse) {
+        allArgUsesCompatible = false;
+        break;
+      }
+    }
```

Comment:
I think it would look cleaner if we use `bool allArgUsesCompatible = llvm::all_of` here. In this case, we can eliminate the early exit in the loop.

Reply:
good idea!